### ISSUE TITLE
Execution of Reduction Operations with TILE SIZE on the GPU in CUDA mode.

### ISSUE THREAD

**Issue by eswarnitk** [libocca/occa#534]
We are trying to run the following code(saved in .okl format) snippet on GPU in CUDA mode.

File : Sample.okl

#define TILE_SIZE_a 8
#define TILE_SIZE_b 8
#define TILE_SIZE_c 4

@kernel void Sample(....)

               for (int ic = cs; ic <= ce; ic++; @tile(TILE_SIZE_c, @outer))
              {
                      for (int ib = bs; ib <= be; ib++; @tile(TILE_SIZE_b, @inner))
                     {
                              int index = (na*nb*ic)+(ib*na);

                              for (int ia = as; ia <= ae; ia++; @tile(TILE_SIZE_a, @inner))
                             {
                                       //arr[index + ia] = z;

                                       arr[index + ia] += z;
                             }
                     }
              }

In the above code, An assignment operator (=) works correctly irrespective of tile sizes. But for the reduction operations ( Ex : +=,-=,*=,/=), it gives wrong output when we use TILE SIZE is less than the loop count. If we take the TILE SIZE is same or more than the loop count, then the program runs successfully.

Ex: If loop count is 64 for all X, Y, and Z loops, if we use 64 as TILE SIZE for all loops, then the reduction operators will perform correctly, but it exceeds the max thread count (1024 < (64*64*64)) in GPU machine.

Is there a dependency on loop count for TILE SIZE for reduction operations?. Can we use the loop count as TILE SIZE?. May i know any other possible solutions for the above code?

Can anyone suggest me how can we run the above code on the GPU in CUDA mode and OKL as our kernel file for any TILE SIZE values?.

Thank you in advance.

Regards,

G Eswar

**Comment by noelchalmers** [comment]
I don't see anything odd about the generated CUDA code for this example
```cpp 
extern "C" __global__ void _occa_Sample_0(const int as,
                                            const int ae,
                                            const int na,
                                            const int bs,
                                            const int be,
                                            const int nb,
                                            const int cs,
                                            const int ce,
                                            const int nc,
                                            const float z,
                                            float * arr) {
  {
    int _occa_tiled_ic = cs + (4 * blockIdx.x);
    for (int ic = _occa_tiled_ic; ic <= (_occa_tiled_ic + 4); ic++) {
      if (ic <= ce) {
        {
          int _occa_tiled_ib = bs + (8 * threadIdx.y);
          for (int ib = _occa_tiled_ib; ib <= (_occa_tiled_ib + 8); ib++) {
            if (ib <= be) {
              int index = (na * nb * ic) + (ib * na);
              {
                int _occa_tiled_ia = as + (8 * threadIdx.x);
                for (int ia = _occa_tiled_ia; ia <= (_occa_tiled_ia + 8); ia++) {
                  if (ia <= ae) {
                    //arr[index + ia] = z;
                    arr[index + ia] += z;
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```
Provided each of the `n<x>` variables are proper counts, i.e. `na == ae-as+1`, there are no data races or anything of the sort.

Considering you say the assignment works, I recommend you check your code for indexing problems. 

P.S. I'm not sure what about this example is a 'reduction'. This looks like it's just adding a scalar to each entry of vector. 

P.P.S. Note that since the tiled loop for `ic` is an `@outer` loop, its tile size has nothing to do with the max threads per block. `@outer` loops are mapped to blocks in the grid. This means something like `TILE_SIZE_a=TILE_SIZE_b=TILE_SIZE_c=32` works fine since `TILE_SIZE_a*TILE_SIZE_b=1024`.

**Comment by eswarnitk** [comment]
Thanks for the quick reply. Yes you are right, even i noticed that there is no problem with the code generation. It is printing the index values correctly, but for the "+=" operation the values are wrong in the final array.
Ex: Initially, my array values are 0.0 and tried to add 2.0 to the array with += operation, so at the end all values in the final array should be 2.0. If i set TILE SIZE is less than the loop count, then the values are coming incorrect. I get the correct values if i set the TILE SIZE is more than the loop count for all three loops.

**Comment by noelchalmers** [comment]
Gotcha. I think I may see the problem, and it has to do with the `<=` in the `for (int ic = _occa_tiled_ic; ic <= (_occa_tiled_ic + 4); ic++) {` for loops. 

Does 
```cpp
 for (int ic = cs; ic < ce+1; ic++; @tile(TILE_SIZE_c, @outer))  {
   for (int ib = bs; ib < be+1; ib++; @tile(TILE_SIZE_b, @inner)) {
     int index = (na*nb*ic)+(ib*na);

     for (int ia = as; ia < ae+1; ia++; @tile(TILE_SIZE_a, @inner)) {
       arr[index + ia] += z;
     }
  }
}
```
work? (Note the slight change in the loops' upper bounds).

**Comment by aroobam** [comment]
Bingo Noel. Man, we just spent the last few hours getting there and when I got here to post that "finding" I see your reply waiting for us :-). Yes, when we change the "<=" to "<" n + 1 things work and all combinations of tile sizes work. We are guessing some border overlap in defining smaller blocks using tile sizes?

Also we pulled out the latest libocca and saw now we are able to do printf() from within an okl file, even though it complains during JIT time. This helped us narrow down the above behavior.

BTW, I work with Eswar, the author of this post. 

Thanks and Regards,
Madhan.
